tcms/rpc/api/testexecution.py

```diff
@@ -183,12 +183,27 @@ def update(execution_id, values, **kwargs):
 
     # if this call updated TE.status then adjust timestamps
     if values.get("status"):
+        now = timezone.now()
         if test_execution.status.weight != 0:
-            test_execution.close_date = timezone.now()
+            test_execution.close_date = now
         else:
             test_execution.close_date = None
         test_execution.save()
 
+        all_runs = TestExecution.objects.filter(run=test_execution.run)
+        if test_execution.status.weight > 0:
+            all_positive = not all_runs.filter(status__weight__lt=0).exists()
+            if all_positive:
+                run = TestRun.objects.get(pk=test_execution.run.pk)
+                run.stop_date = now
+                run.save()
+        elif test_execution.status.weight <= 0:
+            all_negative = not all_runs.filter(status__weight__gt=0).exists()
+            if all_negative:
+                run = TestRun.objects.get(pk=test_execution.run.pk)
+                run.stop_date = None
+                run.save()
```

Comment:
This comparison is wrong. Both Negative and Positive statuses are considered complete. Only weight == 0 is "in progress".